presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java

```diff
@@ -87,68 +90,101 @@ public void addSystemSessionProperties(List<PropertyMetadata<?>> systemSessionPr
         requireNonNull(sessionProperties, "sessionProperties is null");
 
-        checkState(catalogSessionProperties.putIfAbsent(catalog, sessionProperties) == null, "SessionProperties for catalog '%s' are already registered", catalog);
-        for (PropertyMetadata<?> sessionProperty : sessionProperties) {
-            addSessionPropertyInternal(Optional.of(catalog), sessionProperty);
+        List<SessionPropertyMetadata<?>> properties = sessionProperties.stream()
+                .map(sessionProperty -> new SessionPropertyMetadata<>(Optional.of(catalog), Optional.of(connectorId), sessionProperty))
+                .collect(toList());
+
+        Map<String, SessionPropertyMetadata<?>> propertiesByName = Maps.uniqueIndex(properties, SessionPropertyMetadata::getPropertyName);
+        checkState(connectorSessionProperties.putIfAbsent(connectorId, propertiesByName) == null, "SessionProperties for connectorId '%s' are already registered", connectorId);
+
+        for (SessionPropertyMetadata<?> property : properties) {
+            allSessionProperties.put(property.getFullyQualifiedName(), property);
         }
     }
 
-    private <T> void addSessionPropertyInternal(Optional<String> catalogName, PropertyMetadata<T> sessionProperty)
+    public Optional<ConnectorId> getSessionPropertyConnectorId(String name)
     {
-        SessionProperty<T> value = new SessionProperty<>(catalogName, sessionProperty);
-        allSessionProperties.put(value.getFullyQualifiedName(), value);
+        requireNonNull(name, "name is null");
+
+        SessionPropertyMetadata<?> sessionPropertyMetadata = allSessionProperties.get(name);
+        if (sessionPropertyMetadata == null) {
+            throw new PrestoException(INVALID_SESSION_PROPERTY, "Unknown session property " + name);
+        }
+        return sessionPropertyMetadata.getConnectorId();
     }
 
-    public PropertyMetadata<?> getSessionPropertyMetadata(String name)
+    public SessionPropertyMetadata<?> getSessionPropertyMetadata(String name)
     {
         requireNonNull(name, "name is null");
 
-        SessionProperty<?> sessionProperty = allSessionProperties.get(name);
-        if (sessionProperty == null) {
+        SessionPropertyMetadata<?> sessionPropertyMetadata = allSessionProperties.get(name);
+        if (sessionPropertyMetadata == null) {
             throw new PrestoException(INVALID_SESSION_PROPERTY, "Unknown session property " + name);
         }
-        return sessionProperty.getMetadata();
+        return sessionPropertyMetadata;
+    }
+
+    public SessionPropertyMetadata<?> getSessionPropertyMetadata(ConnectorId connectorId, String propertyName)
+    {
+        requireNonNull(connectorId, "connectorId is null");
+        requireNonNull(propertyName, "propertyName is null");
+        Map<String, SessionPropertyMetadata<?>> properties = connectorSessionProperties.get(connectorId);
+        if (properties == null || properties.isEmpty()) {
```

Comment:
Is this `isEmpty()` check correct? The `addConnectorSessionProperties()` method doesn't validate that the list of properties is non-empty.